panther_driver/src/driver_node.py

```diff
@@ -317,49 +329,85 @@ def driverNode():
+                wheel_pos[2] = rear_controller.sdo['Qry_ABSPEED'][2].raw
             except:
                 rospy.logwarn("Error reading rear left controller sdo")
             try:  
-                position_RR = rear_controller.sdo['Qry_ABCNTR'][1].raw
+                wheel_pos[3] = rear_controller.sdo['Qry_ABSPEED'][1].raw
             except:
                 rospy.logwarn("Error reading rear right controller sdo")
 
-            # position_FL / RK.encoder_resolution - > full wheel rotations
-            wheel_FL_ang_pos = 2 * math.pi * position_FL / RK.encoder_resolution  # radians
-            wheel_FR_ang_pos = 2 * math.pi * position_FR / RK.encoder_resolution
-            wheel_RL_ang_pos = 2 * math.pi * position_RL / RK.encoder_resolution
-            wheel_RR_ang_pos = 2 * math.pi * position_RR / RK.encoder_resolution
 
-            wheel_FL_ang_vel = (wheel_FL_ang_pos -
-                                wheel_FL_ang_pos_last) * dt_  # rad/s
-            wheel_FR_ang_vel = (wheel_FR_ang_pos - wheel_FR_ang_pos_last) * dt_
-            wheel_RL_ang_vel = (wheel_RL_ang_pos - wheel_RL_ang_pos_last) * dt_
-            wheel_RR_ang_vel = (wheel_RR_ang_pos - wheel_RR_ang_pos_last) * dt_
+            # query current
+            # division by 10 is needed according to documentation
+            try:
+                wheel_curr[0] = front_controller.sdo['Qry_MOTAMPS'][2].raw / 10.0
+            except:
+                rospy.logwarn("Error reading current front left controller sdo")
+            try:
+                wheel_curr[1] = front_controller.sdo['Qry_MOTAMPS'][1].raw / 10.0
+            except:
+                rospy.logwarn("Error reading current front right controller sdo")
+            try:
+                wheel_curr[2] = rear_controller.sdo['Qry_MOTAMPS'][2].raw / 10.0
+            except:
+                rospy.logwarn("Error reading current rear left controller sdo")
+            try:  
+                wheel_curr[3] = rear_controller.sdo['Qry_MOTAMPS'][1].raw / 10.0
+            except:
+                rospy.logwarn("Error reading current rear right controller sdo")
+
+
+            joint_state_msg.header.stamp = rospy.Time.now()
+            joint_state_msg.header.seq = seq_counter
 
-            wheel_FL_ang_pos_last = wheel_FL_ang_pos
-            wheel_FR_ang_pos_last = wheel_FR_ang_pos
-            wheel_RL_ang_pos_last = wheel_RL_ang_pos
-            wheel_RR_ang_pos_last = wheel_RR_ang_pos
+
+            # convert tics to rad
+            joint_state_msg.position = [2.0 * math.pi * pos / RK.encoder_resolution / RK.gear_ratio for pos in wheel_pos]
+            # convert RPM to rad/s
+            joint_state_msg.velocity = [vel / 60.0 * 2.0 * math.pi / RK.gear_ratio for vel in wheel_vel]
+            # convert to A to Nm
+            joint_state_msg.effort = [
+               wheel_curr[i] * motor_torque_constant * sign(wheel_vel[i]) for i in range(wheel_curr)]
```

Comment:
what if `wheel_vel[i]` is 0? I think that both your `sign` function and the one from numpy will return 0. Should effort output be 0 then?

Reply:
Ok. I will update that sign function to either be 1 or -1. In theory, float value should never be equal to 0, but I will cover that edgecase.